core/vm/jit.go

```diff
@@ -0,0 +1,592 @@
+// Copyright 2014 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package vm
+
+import (
+	"fmt"
+	"math/big"
+	"sync/atomic"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/state"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/params"
+	"github.com/hashicorp/golang-lru"
+)
+
+type progStatus int32
+
+const (
+	progUnknown progStatus = iota
+	progCompile
+	progReady
+	progError
+)
+
+var programs *lru.Cache
+
+func init() {
+	programs, _ = lru.New(defaultJitMaxCache)
+}
+
+// SetJITCacheSize recreates the program cache with the max given size. Setting
+// a new cache is **not** thread safe. Use with caution.
+func SetJITCacheSize(size int) {
+	programs, _ = lru.New(size)
+}
+
+// GetProgram returns the program by id or nil when non-existent
+func GetProgram(id common.Hash) *Program {
+	if p, ok := programs.Get(id); ok {
+		return p.(*Program)
+	}
+
+	return nil
+}
+
+// GenProgramStatus returns the status of the given program id
+func GetProgramStatus(id common.Hash) progStatus {
+	program := GetProgram(id)
+	if program != nil {
+		return progStatus(atomic.LoadInt32(&program.status))
+	}
+
+	return progUnknown
+}
+
+// Program is a compiled program for the JIT VM and holds all required for
+// running a compiled JIT program.
+type Program struct {
+	Id     common.Hash // Id of the program
+	status int32       // status should be accessed atomically
+
+	context *Context
+
+	instructions []instruction       // instruction set
+	mapping      map[uint64]int      // real PC mapping to array indices
+	destinations map[uint64]struct{} // cached jump destinations
+
+	code []byte
+}
+
+func NewProgram(code []byte) *Program {
+	program := &Program{
+		Id:           crypto.Sha3Hash(code),
```

Comment:
Hmm, we could potentially avoid this hashing by reusing the contract address as id, but then it's not unique on code. What do you think?